draft-ietf-tls-tls13.md

```diff
@@ -5636,6 +5762,69 @@ application protocols separately ensuring that confidential
+actions may cause side effects (e.g., purchasing an item or
+transferring money) to be duplicated, thus harming the site or the
+user.  Attackers can also store and replay 0-RTT messages in order to
+re-order them with respect to other messages, which may harm
+unsuspecting applications.
+
+In addition, if data can be replayed a large number of times,
+additional attacks become possible. Specifically, attackers
+can use multiple replays to exploit information leakage via
+side channels such as timing network caches or measuring the
+speed of cryptographic. In addition, they may be able to
+overload rate-limiting systems. For further description of
+these attacks, see {{Mac17}}.
+
+Ultimately, servers have the responsibility to protect themselves
+against attacks employing 0-RTT data replication. The mechanism
+described in {{anti-replay}} are intended to
+prevent replay at the TLS layer do not provide complete protection
+against receiving multiple copies of client data:
+TLS 1.3 falls back to the 1-RTT
+handshake when the server does not have any information about the
+client, e.g., because it is in a different cluster which does not
+share state or because the ticket has been deleted as described in
+{{single-use-tickets}}. If the application layer protocol retransmits
+data in this setting, then it is possible for an attacker to induce a
+replay attack by sending the ClientHello to both the original cluster
+(which processes the data immediately) and another cluster which will
+fall back to 1-RTT and process the data upon application layer
+replay. The scale of this attack is limited by the client's
+willingness to replay and therefore only allows a limited number of
+replays, which will also use different encryption keys.
+
+If implemented correctly, the mechanisms described in
+{{single-use-tickets}} and {{client-hello-recording}}, prevent a
+replayed ClientHello and its associated 0-RTT data from being accepted
+multiple times by any cluster with consistent state; for server
+which limit the use of 0-RTT to one cluster for a single ticket, then a given
+ClientHello and its associated 0-RTT data will only be accepted once.
+However, if state is not completely consistent,
+then an attacker might be able to have multiple copies of the data be
+accepted during the replication window.
+
+Because they do not know the exact details of server behavior,
+clients MUST NOT send
+messages in early data which are not safe to have replayed and which
+they would not be willing to retry across multiple 1-RTT
+connections. Protocols MUST NOT use 0-RTT data without a profile that
+defines its use. That profile needs to identify which messages or
+interactions are safe to use with 0-RTT. In addition, to avoid
+accidental misuse, implementations MUST NOT enable 0-RTT (either
+sending or receiving) unless specifically requested by the application
+and SHOULD NOT automatically resend 0-RTT data if it is rejected by
```

Comment:
Not sure how to manage this SHOULD NOT.  You should definitely qualify the "reject" here to be "if early data is rejected by the server"; the application protocol can (and SHOULD) have ways to signal that requests sent in early data are not safe.

The HTTP profile we just published requires automatic retries of requests sent in early data in clear violation of this SHOULD.  I see no sensible alternative.  Once you have accepted the risk of replay, then a single retry isn't fatal.

My suggestion: Force the profile to specify what happens when early data is rejected, but don't say anything about it here.